Solve Issue #63: Add Support for Token Authentication and User/Pass Authentication in BMC Firmware Update

abhashsolanki18 opened the pull request:
Issue Addressed: Resolves #63
Changes Made:
1. Added support for token-based authentication alongside existing user/password authentication in the BMC firmware update process.
2. Introduced two distinct blocks in roles/bmc_fw_update/tasks/main.yml:
User/Password Authentication Block: For handling authentication using username and password.
Token Authentication Block: For handling authentication using tokens.

Details:
Updated the community.general.redfish_info task to support both authentication methods.
Ensured compatibility with existing configurations and backward compatibility.

@glimchb  Please review these changes. Your feedback is appreciated!

Commit:
fix: support-token-user/pass

Signed-off-by: abhash abhashsolanki18@gmail.com
Signed-off-by: root <root@localhost.localdomain>

glimchb commented:
@abhashsolanki18 thanks for fixing this
Please see commitlint failure.
Otherwise looks good.
If you rebase, those ansible ci failures should be resolved

abhashsolanki18 commented:
@glimchb i have rebased the commits now it should pass the ci.

glimchb commented on roles/bmc_fw_update/tasks/main.yml:
```diff
@@ -113,10 +171,6 @@
     category: Update
     command: GetUpdateStatus
     baseuri: "{{ inventory_hostname }}"
-    username: "{{ dpu_bmc_username }}"
```
@abhashsolanki18 why this is removed ?

glimchb replied:
@abhashsolanki18 should this be either `username` and `password` or `auth_token` ?
OR
can we apply ` | default(omit)`  everywhere above instead of creating new block ? 

WDYT ?

abhashsolanki18 replied:
@glimchb 
applied default (omit) everywhere that should work and simplify the code by removing the blocks.
see the latest commit

glimchb replied:
ok, make sure you test it works... I had issues with this before...

abhashsolanki18 replied:
it's working perfectly for me

glimchb commented:
@abhashsolanki18 
- looks like linter fails on your changes, please fix
- i saw one place at the end when you don;t give neither user/pass nor auth token... looks like a bug...

glimchb reviewed (changes requested):
@abhashsolanki18 
- looks like linter fails on your changes, please fix
- i saw one place at the end when you don;t give neither user/pass nor auth token... looks like a bug...

abhashsolanki18 committed:
linters fix

Signed-off-by: Abhash Solanki <abhashsolanki18@gmail.com>

abhashsolanki18 commented:
@glimchb 
the linters should also exclude ansible 2.17 with python 3.9, fixed that in the latest commit
also fixed the bug you highlighted earlier.

glimchb commented on .github/workflows/ansible.yml:
```diff
@@ -104,13 +104,14 @@ jobs:
-        python-version: ["3.9", "3.10", "3.11"]
-        ansible-version: [stable-2.17, stable-2.15, stable-2.16]
-        exclude:
-        # Ansible-core 2.16 is supported only from Python 3.10 onwards
```
maybe better to use `["3.10", "3.11", "3.12"]` in the matrix and remove those excludes?

abhashsolanki18 replied:
yes, maybe that would be better. changed it int the latest commit

glimchb commented on roles/bmc_fw_update/tasks/main.yml:
```diff
@@ -154,4 +212,5 @@
   when:
     - bmc_fw_update_reboot is true
-    - not bmc_fw_update_image_file is search(bmc_fw_update_got_fw_version | regex_search('[0-9-.]+'))
+    - not bmc_fw_update_image_file is search(bmc_fw_update_got_fw_version | regex_search('[0-9-.]+'))"
```
last `"` is typo?

abhashsolanki18 replied:
yes, fixed it

abhashsolanki18 committed:
fix(bmc_fw_update): resolve YAML linting issues

This commit resolves multiple YAML linting issues in the `bmc_fw_update` task file.
- Removed extra blank lines
- Ensured all tasks are named
- Improved task key order
- Prefixed variable names appropriately

These changes should ensure the playbook passes `yamllint` and `ansible-lint` checks.

Signed-off-by: Abhash Solanki <abhashsolanki18@gmail.com>

abhashsolanki18 commented:
[Invalid workflow file: .github/workflows/docker-publish.yml#L13](https://github.com/abhashsolanki18/ansible-opi-dpu/actions/runs/9902996175/workflow)
The workflow is not valid. .github/workflows/docker-publish.yml (Line: 13, Col: 3): Error calling workflow 'opiproject/actions/.github/workflows/docker-publish.yml@main'. The nested job 'build' is requesting 'id-token: write', but is only allowed 'id-token: none'.

@glimchb what is the error about? failing while executing docker workflow as well as openSSF workflow

glimchb commented on .github/workflows/ansible.yml:
```diff
@@ -104,13 +99,14 @@ jobs:
-        - python-version: "3.9"
-          ansible-version: stable-2.16
-
+         python-version: ["3.9", "3.10", "3.11"]
```
```suggestion
         python-version: ["3.10", "3.11", "3.12"]
```

glimchb commented on .github/workflows/ansible.yml:
```diff
@@ -104,13 +99,14 @@ jobs:
+          # Ansible-core 2.16 and 2.17 are supported only from Python 3.10 onwards
+          - python-version: "3.9"
+            ansible-version: stable-2.16
+          - python-version: "3.9"
```
why not like above just remove 3.9

abhashsolanki18 replied:
done in the latest commit

glimchb commented on roles/bmc_fw_update/tasks/main.yml:
```diff
@@ -43,61 +43,62 @@
-    username: "{{ dpu_bmc_username }}"
-    password: "{{ dpu_bmc_password }}"
-    # auth_token: "{{ dpu_bmc_token }}"
+    auth_token: "{{ dpu_bmc_token | default(omit) }}"
```
did you test this ? I think this failed for me last time I tried it...

glimchb commented:
> @glimchb what is the error about? failing while executing docker workflow as well as openSSF workflow


will check it

abhashsolanki18 committed:
Merge branch 'opiproject:main' into main

abhashsolanki18 committed:
fix(bmc_fw_update): resolve workflow permission errors

Signed-off-by: Abhash Solanki <abhashsolanki18@gmail.com>

abhashsolanki18 committed:
fix(bmc_fw_update): resolve workflow permission errors

Signed-off-by: Abhash Solanki <abhashsolanki18@gmail.com>

glimchb commented on roles/bmc_fw_update/tasks/main.yml:
```diff
@@ -23,11 +23,11 @@
 - name: Store current fw version
   ansible.builtin.set_fact:
-    bmc_fw_update_cur_fw_version: "{{ vars.get_bmc_facts_all_fw_versions[bmc_fw_update_inventory_name] }}"
+    bmc_fw_update_cur_fw_version: "{{ get_bmc_facts_before.get_bmc_facts_all_fw_versions[bmc_fw_update_inventory_name] }}"
```
this doesn't work in my setup for some reason, let's try to understand why?

glimchb commented on roles/bmc_fw_update/tasks/main.yml:
```diff
@@ -23,11 +23,11 @@
 - name: Print BMC Version
   ansible.builtin.debug:
-    msg: "{{ get_bmc_facts_all_fw_versions }}"
+    msg: "{{ get_bmc_facts_before.get_bmc_facts_all_fw_versions }}"
```
this doesn't work in my setup for some reason, let's try to understand why?

glimchb commented:
@abhashsolanki18 I tested your changes and they fail, please fix

I used for user/pass flow:
```
ansible-playbook ./playbooks/firmware.yml -vvv -i "172.22.4.2," -e dpu_bmc_username='root' -e dpu_bmc_password='123456' -e bmc_fw_update_inventory_name='6d53cf4d_bmc_firmware' -e bmc_fw_update_image_file='/tmp/bf2-bmc-ota-24.04-5-opn.tar'
```
I hit errors, so I had to make those changes:
```diff
diff --git a/roles/bmc_fw_update/defaults/main.yml b/roles/bmc_fw_update/defaults/main.yml
index ba5ffaf..ba8a611 100644
--- a/roles/bmc_fw_update/defaults/main.yml
+++ b/roles/bmc_fw_update/defaults/main.yml
@@ -8,5 +8,5 @@ bmc_fw_update_reboot: true
 bmc_fw_update_job_wait: true
 bmc_fw_update_inventory_name: bmc_firmware
 bmc_fw_update_image_file: /tmp/bf3-bmc-24.04-5_opn.fwpkg
-bmc_fw_update_image_url: https://content.mellanox.com/BlueField/BMC/24.04-5-Apr-2024/{{ bmc_fw_update_image_file | basename }}
+bmc_fw_update_image_url: https://www.mellanox.com/downloads/BlueField/BMC/24.04-5-Apr-2024/{{ bmc_fw_update_image_file | basename }}
 https_port: 443
diff --git a/roles/bmc_fw_update/tasks/main.yml b/roles/bmc_fw_update/tasks/main.yml
index e37647c..93b8d05 100644
--- a/roles/bmc_fw_update/tasks/main.yml
+++ b/roles/bmc_fw_update/tasks/main.yml
@@ -23,11 +23,11 @@

 - name: Store current fw version
   ansible.builtin.set_fact:
-    bmc_fw_update_cur_fw_version: "{{ get_bmc_facts_before.get_bmc_facts_all_fw_versions[bmc_fw_update_inventory_name] }}"
+    bmc_fw_update_cur_fw_version: "{{ vars.get_bmc_facts_all_fw_versions[bmc_fw_update_inventory_name] }}"

 - name: Print BMC Version
   ansible.builtin.debug:
-    msg: "{{ get_bmc_facts_before.get_bmc_facts_all_fw_versions }}"
+    msg: "{{ get_bmc_facts_all_fw_versions }}"

 - name: Check if firmware image exists locally {{ bmc_fw_update_image_file }}
   ansible.builtin.stat:
```

once fixed, I can test the same with token:

```
ansible-playbook ./playbooks/firmware.yml -vvv -i "172.22.4.2," -e dpu_bmc_token='123456' -e bmc_fw_update_inventory_name='6d53cf4d_bmc_firmware' -e bmc_fw_update_image_file='/tmp/bf2-bmc-ota-24.04-5-opn.tar'
```

glimchb commented on roles/bmc_fw_update/tasks/main.yml:
```diff
@@ -143,15 +143,16 @@
 
 - name: Store fw version we installed
   ansible.builtin.set_fact:
-    bmc_fw_update_got_fw_version: "{{ vars.get_bmc_facts_all_fw_versions[bmc_fw_update_inventory_name] }}"
```
this doesn't work in my setup for some reason, let's try to understand why?

abhashsolanki18 commented:
> @abhashsolanki18 I tested your changes and they fail, please fix
> 
> I used for user/pass flow:
> 
> ```
> docker run --rm -it --entrypoint ansible-playbook -v $(pwd):/opt -w /opt/roles ghcr.io/opiproject/ansible-opi-dpu:main ../playbooks/firmware.yml -vvv -i "172.22.4.2," -e dpu_bmc_username='root' -e dpu_bmc_password='123456' -e bmc_fw_update_inventory_name='6d53cf4d_bmc_firmware' -e bmc_fw_update_image_file='/tmp/bf2-bmc-ota-24.04-5-opn.tar'
> ```
> 
> I hit errors, so I had to make those changes:
> 
> ```diff
> diff --git a/roles/bmc_fw_update/defaults/main.yml b/roles/bmc_fw_update/defaults/main.yml
> index ba5ffaf..ba8a611 100644
> --- a/roles/bmc_fw_update/defaults/main.yml
> +++ b/roles/bmc_fw_update/defaults/main.yml
> @@ -8,5 +8,5 @@ bmc_fw_update_reboot: true
>  bmc_fw_update_job_wait: true
>  bmc_fw_update_inventory_name: bmc_firmware
>  bmc_fw_update_image_file: /tmp/bf3-bmc-24.04-5_opn.fwpkg
> -bmc_fw_update_image_url: https://content.mellanox.com/BlueField/BMC/24.04-5-Apr-2024/{{ bmc_fw_update_image_file | basename }}
> +bmc_fw_update_image_url: https://www.mellanox.com/downloads/BlueField/BMC/24.04-5-Apr-2024/{{ bmc_fw_update_image_file | basename }}
>  https_port: 443
> diff --git a/roles/bmc_fw_update/tasks/main.yml b/roles/bmc_fw_update/tasks/main.yml
> index e37647c..93b8d05 100644
> --- a/roles/bmc_fw_update/tasks/main.yml
> +++ b/roles/bmc_fw_update/tasks/main.yml
> @@ -23,11 +23,11 @@
> 
>  - name: Store current fw version
>    ansible.builtin.set_fact:
> -    bmc_fw_update_cur_fw_version: "{{ get_bmc_facts_before.get_bmc_facts_all_fw_versions[bmc_fw_update_inventory_name] }}"
> +    bmc_fw_update_cur_fw_version: "{{ vars.get_bmc_facts_all_fw_versions[bmc_fw_update_inventory_name] }}"
> 
>  - name: Print BMC Version
>    ansible.builtin.debug:
> -    msg: "{{ get_bmc_facts_before.get_bmc_facts_all_fw_versions }}"
> +    msg: "{{ get_bmc_facts_all_fw_versions }}"
> 
>  - name: Check if firmware image exists locally {{ bmc_fw_update_image_file }}
>    ansible.builtin.stat:
> ```
> 
> once fixed, I can test the same with token:
> 
> ```
> curl -k -H "Content-Type: application/json" -X POST https://172.22.4.2/login -d '{"username" : "root", "password" : "123456"}'
> docker run --rm -it --entrypoint ansible-playbook -v $(pwd):/opt -w /opt/roles ghcr.io/opiproject/ansible-opi-dpu:main ../playbooks/firmware.yml -vvv -i "172.22.4.2," -e dpu_bmc_token='123456' -e bmc_fw_update_inventory_name='6d53cf4d_bmc_firmware' -e bmc_fw_update_image_file='/tmp/bf2-bmc-ota-24.04-5-opn.tar'    
> ```

will find out whats going wrong
